src/did.ts

```diff
@@ -0,0 +1,321 @@
+   */
+  setResolver(resolver: Resolver | ResolverRegistry, resolverOptions?: ResolverOptions): void {
+    this._resolver = isResolver(resolver) ? resolver : new Resolver(resolver, resolverOptions)
+  }
+
+  /**
+   * Authenticate the user.
+   */
+  async authenticate({ provider, paths = [], aud }: AuthenticateOptions = {}): Promise<string> {
+    if (provider != null) {
+      this.setProvider(provider)
+    }
+    if (this._client == null) {
+      throw new Error('No provider available')
+    }
+    const nonce = randomString()
+    const jws = await this._client.request('did_authenticate', {
+      nonce,
+      aud,
+      paths,
+    })
+    const { kid } = await this.verifyJWS(jws)
+    const payload = base64urlToJSON(jws.payload) as AuthenticateResponse
+    if (!kid.includes(payload.did)) throw new Error('Invalid authencation response, kid mismatch')
+    if (payload.nonce !== nonce) throw new Error('Invalid authencation response, wrong nonce')
+    if (payload.aud !== aud) throw new Error('Invalid authencation response, wrong aud')
+    if (payload.exp < Date.now() / 1000) throw new Error('Invalid authencation response, expired')
+    this._id = payload.did
+    return this._id
+  }
+
+  /**
+   * Create a JWS encoded signature over the given payload.
+   * Will be signed by the currently authenticated DID.
+   *
+   * @param payload             The payload to sign
+   * @param options             Optional parameters
+   */
+  async createJWS<T = any>(payload: T, options: CreateJWSOptions = {}): Promise<DagJWS> {
+    if (this._client == null) throw new Error('No provider available')
+    if (this._id == null) throw new Error('DID is not authenticated')
+    const { jws } = await this._client.request('did_createJWS', {
+      did: this._id,
+      ...options,
+      payload,
+    })
+    return jws
+  }
+
+  /**
+   * Create an IPFS compatibe DagJWS encoded signature over the given payload.
+   * Will be signed by the currently authenticated DID.
+   *
+   * @param payload             The payload to sign, may include ipld links
+   * @param options             Optional parameters
+   */
+  async createDagJWS(
+    payload: Record<string, any>,
+    options: CreateJWSOptions = {}
+  ): Promise<DagJWSResult> {
+    const { cid, linkedBlock } = await encodePayload(payload)
+    const payloadCid = encodeBase64Url(cid.bytes)
+    Object.assign(options, { linkedBlock: encodeBase64(linkedBlock) })
+    const jws = await this.createJWS(payloadCid, options)
+    jws.link = cid
+    return { jws, linkedBlock }
+  }
+
+  /**
+   * Verify a JWS. Uses the 'kid' in the header as the way to resolve
+   * the author public key.
+   *
+   * @param jws                 The JWS to verify
+   * @param atTime              JS timestamp when to verify the key
+   * @returns                   Information about the signed JWS
+   */
+  async verifyJWS(
+    jws: string | DagJWS,
+    atTime: number = new Date().valueOf()
```

Comment:
I would pass an options object here since that gives us more flexibility if we need to other params in the future.

Reply:
kk